### PR TITLE
Pandera integration

### PR DESCRIPTION
**Pull request by cosmicBboy**
# Pandera Integration

This PR adds a [pandera](https://pandera.readthedocs.io/en/stable/) integration to the `schemas.py` module, allowing users to specify dataframe schemas using the [pandera class-based API](https://pandera.readthedocs.io/en/stable/schema_models.html)

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [X] Plugin

## Are all requirements met?

TODO: testing

 - [X] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

Add `pandera` plugin and register `PanderaTransformer` to the type engine.

Example script:

```python
import pandas as pd
from flytekit import task, workflow
from flytekit.types import pandera as pa
from pandera.typing import DataFrame, Series


class InSchema(pa.SchemaModel):
    hourly_pay: Series[float] = pa.Field(ge=7)
    hours_worked: Series[float] = pa.Field(ge=10)

    class Config:
        coerce = True


class IntermediateSchema(InSchema):
    total_pay: Series[float]

    @pa.dataframe_check(name="foo")
    def check_total_pay(cls, df: DataFrame) -> Series[bool]:
        """Run-time validation check."""
        return df["total_pay"] == df["hourly_pay"] * df["hours_worked"]


class OutSchema(IntermediateSchema):
    worker_id: Series[str]


@task
def total_pay(df: DataFrame[InSchema]) -> DataFrame[IntermediateSchema]:
    return df.assign(total_pay=df.hourly_pay * df.hours_worked)


@task
def add_id(df: DataFrame[IntermediateSchema], id: str) -> DataFrame[OutSchema]:
    return df.assign(worker_id=id)


@workflow
def pandera_workflow() -> DataFrame[OutSchema]:
    worker_id = 100
    df = pd.DataFrame(
        {
            "hourly_pay": ["12.0", "13.5", "10.1"],
            "hours_worked": [30.5, 40.0, 41.75]
        }
    )
    return add_id(df=total_pay(df=df), id=worker_id)


if __name__ == "__main__":
    print(f"Running {__file__} main...")
    result = pandera_workflow()
    print(f"Running wf(), returns dataframe\n{result}\n{result.dtypes}")

```

Output:
```python
Running example.py main...
Running wf(), returns dataframe
   hourly_pay  hours_worked  total_pay worker_id
0        12.0         30.50    366.000       100
1        13.5         40.00    540.000       100
2        10.1         41.75    421.675       100
hourly_pay      float64
hours_worked    float64
total_pay       float64
worker_id        object
dtype: object
```

## Tracking Issue

TODO: make an issue for this PR

https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by kumare3** [comment]
@cosmicBboy you may want to take my latest PR, cleans up a lot of the concepts around schema. https://github.com/lyft/flytekit/pull/267/files

**Comment by cosmicBboy** [comment]
Cool, I'll take a look and wait 'til those changes are merged

**Comment by cosmicBboy** [comment]
just rebased changes onto the up-to-date `annotations` branch, and added a few tests. Not sure how to organize the plugin, for now it's being imported in `types/__init__.py` so users can access the `pandera` package namespace with `from flytekit.types import pandera`, which should also register the `PanderaSchemaTransformer` on the type engine.

**Review by wild-endeavor — APPROVED** [review]
thanks!

**Comment by wild-endeavor** [comment]
I don't have permissions to merge anymore (getting them back once it moves to linux foundation).  @anandswaminathan please?

**Review by anandswaminathan — APPROVED** [review]

File: setup.py [hunk 1]
```diff
@@ -29,6 +29,7 @@
         "k8s-proto>=0.0.3,<1.0.0",
         "numpy>=1.14.0,<2.0.0",
         "pandas>=0.22.0,<2.0.0",
+        "pandera>=0.5.0",
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply setup.py:32]: done!

**Comment by EngHabu** [comment]
Hey @cosmicBboy, this is awesome! Thank you for making this work. In my mind, however, such plugins (type, tasks... etc.) should ideally live closer to the origin (or generally outside of FlyteKit)... in this case, should this integration live in Pandera codebase? so that whoever imports flyte kit and Pandera (so they are aware of both), they can use it... while if they are just writing plain panda dataframes or whatever, they don't need to pull in Pandera library as well... Thoughts?

**Comment by cosmicBboy** [comment]
Hey @EngHabu, that's a good question, based on my conversation with @kumare3 and @wild-endeavor my impression was that pandera would be integrated into the flytekit typing system, so that's where I started for this proof-of-concept.

I haven't delved into the flytekit codebase too much, but how does the plugin system work and are there conventions around how to write them? It looks like currently there are notions of "schemas" for `spark`, `pandas`, and the `FlyteSchema` class, which applies to pandas dataframes (?). As `pandera` is a run-time validation tool for pandas, I just figured it would make sense to have it in the `pandas_schema.py` module as an alternative syntax for expressing pandas schemas.

> whoever imports flyte kit and Pandera (so they are aware of both), they can use it... while if they are just writing plain panda dataframes or whatever, they don't need to pull in Pandera library as well

This would be ideal, but I didn't immediately find a clever way of registering the `PanderaTransformer` into the type engine without having to explicitly import `pandera` from `flytekit.types`. As flytekit is out of my domain of knowledge I'd defer to the flytekit team as to how to achieve this.

**Comment by kumare3** [comment]
@cosmicBboy how are the checks run? Are they run automatically? (I don't think so - but just checking)

**Comment by cosmicBboy** [comment]
@kumare3 yes, validation checks are run automatically in the `to_python_value` method.

also, need to close this PR, since the pandera integration was merged here: https://github.com/flyteorg/flytekit/pull/354

**Pull request closed without merging**